File: go/internal/protocol/protocol.go
package protocol

import (
	"encoding/json"
	"fmt"
	"strconv"
	"strings"
)

type EventKind string

const (
	EventText     EventKind = "TEXT"
	EventToolCall EventKind = "TOOL_CALL"
	EventUsage    EventKind = "USAGE"
	EventStop     EventKind = "STOP"
	EventError    EventKind = "ERROR"
)

type Event interface {
	Kind() EventKind
	Render() string
}

type TextEvent struct{ Content string }

func (e TextEvent) Kind() EventKind { return EventText }
func (e TextEvent) Render() string  { return "TEXT:" + EscapeText(e.Content) }

type StopEvent struct{ Reason string }

func (e StopEvent) Kind() EventKind { return EventStop }
func (e StopEvent) Render() string  { return "STOP:" + e.Reason }

type ErrorEvent struct{ Message string }

func (e ErrorEvent) Kind() EventKind { return EventError }
func (e ErrorEvent) Render() string  { return "ERROR:" + e.Message }

type UsageEvent struct {
	InputTokens      int
	OutputTokens     int
	CacheInputTokens int
}

func (e UsageEvent) Kind() EventKind { return EventUsage }
func (e UsageEvent) Render() string {
	return fmt.Sprintf("USAGE:%d\t%d\t%d", e.InputTokens, e.OutputTokens, e.CacheInputTokens)
}

type ToolCallEvent struct {
	Name      string
	ID        string
	InputJSON json.RawMessage
	Fields    map[string]string
	Order     []string
}

func (e ToolCallEvent) Kind() EventKind { return EventToolCall }
func (e ToolCallEvent) Render() string {
	var b strings.Builder
	b.WriteString("TOOL_CALL:")
	b.WriteString(e.Name)
	b.WriteString("\t")
	b.WriteString(e.ID)
	b.WriteString("\t")
	b.WriteString(EscapeText(string(e.InputJSON)))
	for _, key := range e.Order {
		b.WriteString("\t")
		b.WriteString(key)
		b.WriteString("\t")
		b.WriteString(EscapeText(e.Fields[key]))
	}
	return b.String()
}

func ParseToolCallPayload(payload string) (ToolCallEvent, error) {
	parts := strings.Split(payload, "\t")
	if len(parts) < 3 {
		return ToolCallEvent{}, fmt.Errorf("invalid tool call payload")
	}
	event := ToolCallEvent{
		Name:      parts[0],
		ID:        parts[1],
		InputJSON: json.RawMessage(UnescapeText(parts[2])),
		Fields:    map[string]string{},
	}
	for i := 3; i+1 < len(parts); i += 2 {
		key := parts[i]
		value := UnescapeText(parts[i+1])
		event.Order = append(event.Order, key)
		event.Fields[key] = value
	}
	return event, nil
}

func EscapeText(s string) string {
	s = strings.ReplaceAll(s, "\\", "\\\\")
	s = strings.ReplaceAll(s, "\n", "\\n")
	s = strings.ReplaceAll(s, "\r", "\\r")
	s = strings.ReplaceAll(s, "\t", "\\t")
	return s
}

func UnescapeText(s string) string {
	var b strings.Builder
	b.Grow(len(s))
	for i := 0; i < len(s); i++ {
		if s[i] != '\\' || i+1 >= len(s) {
			b.WriteByte(s[i])
			continue
		}
		i++
		switch s[i] {
		case 'n':
			b.WriteByte('\n')
		case 'r':
			b.WriteByte('\r')
		case 't':
			b.WriteByte('\t')
		case '\\':
			b.WriteByte('\\')
		default:
			b.WriteByte('\\')
			b.WriteByte(s[i])
		}
	}
	return b.String()
}

func ParseUsagePayload(payload string) (UsageEvent, error) {
	parts := strings.Split(payload, "\t")
	if len(parts) != 3 {
		return UsageEvent{}, fmt.Errorf("invalid usage payload")
	}
	in, err := strconv.Atoi(parts[0])
	if err != nil {
		return UsageEvent{}, err
	}
	out, err := strconv.Atoi(parts[1])
	if err != nil {
		return UsageEvent{}, err
	}
	cache, err := strconv.Atoi(parts[2])
	if err != nil {
		return UsageEvent{}, err
	}
	return UsageEvent{InputTokens: in, OutputTokens: out, CacheInputTokens: cache}, nil
}

File: rust/src/protocol.rs
use anyhow::{Result, bail};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::BTreeMap;

#[derive(Debug, Clone)]
pub enum Event {
    Text(TextEvent),
    ToolCall(ToolCallEvent),
    Usage(UsageEvent),
    Stop(StopEvent),
    Error(ErrorEvent),
}

impl Event {
    pub fn render(&self) -> String {
        match self {
            Event::Text(e) => format!("TEXT:{}", escape_text(&e.content)),
            Event::ToolCall(e) => {
                let mut out = format!(
                    "TOOL_CALL:{}\t{}\t{}",
                    e.name,
                    e.id,
                    escape_text(&e.input_json.to_string())
                );
                for key in &e.order {
                    out.push('\t');
                    out.push_str(key);
                    out.push('\t');
                    out.push_str(&escape_text(
                        e.fields.get(key).map(String::as_str).unwrap_or(""),
                    ));
                }
                out
            }
            Event::Usage(e) => format!(
                "USAGE:{}\t{}\t{}",
                e.input_tokens, e.output_tokens, e.cache_input_tokens
            ),
            Event::Stop(e) => format!("STOP:{}", e.reason),
            Event::Error(e) => format!("ERROR:{}", e.message),
        }
    }
}

#[derive(Debug, Clone)]
pub struct TextEvent {
    pub content: String,
}

#[derive(Debug, Clone)]
pub struct StopEvent {
    pub reason: String,
}

#[derive(Debug, Clone)]
pub struct ErrorEvent {
    pub message: String,
}

#[derive(Debug, Clone)]
pub struct UsageEvent {
    pub input_tokens: i64,
    pub output_tokens: i64,
    pub cache_input_tokens: i64,
}

#[derive(Debug, Clone)]
pub struct ToolCallEvent {
    pub name: String,
    pub id: String,
    pub input_json: Value,
    pub fields: BTreeMap<String, String>,
    pub order: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToolResultContent {
    #[serde(rename = "type")]
    pub kind: String,
    pub tool_use_id: String,
    pub content: String,
}

pub fn escape_text(s: &str) -> String {
    s.replace('\\', "\\\\")
        .replace('\n', "\\n")
        .replace('\r', "\\r")
        .replace('\t', "\\t")
}

pub fn unescape_text(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    let mut chars = s.chars();
    while let Some(c) = chars.next() {
        if c != '\\' {
            out.push(c);
            continue;
        }
        match chars.next() {
            Some('n') => out.push('\n'),
            Some('r') => out.push('\r'),
            Some('t') => out.push('\t'),
            Some('\\') => out.push('\\'),
            Some(other) => {
                out.push('\\');
                out.push(other);
            }
            None => out.push('\\'),
        }
    }
    out
}

pub fn parse_tool_call_payload(payload: &str) -> Result<ToolCallEvent> {
    let parts: Vec<&str> = payload.split('\t').collect();
    if parts.len() < 3 {
        bail!("invalid tool call payload")
    }
    let input: Value = serde_json::from_str(&unescape_text(parts[2]))?;
    let mut fields = BTreeMap::new();
    let mut order = Vec::new();
    let mut i = 3;
    while i + 1 < parts.len() {
        let key = parts[i].to_string();
        let value = unescape_text(parts[i + 1]);
        fields.insert(key.clone(), value);
        order.push(key);
        i += 2;
    }
    Ok(ToolCallEvent {
        name: parts[0].to_string(),
        id: parts[1].to_string(),
        input_json: input,
        fields,
        order,
    })
}

pub fn parse_usage_payload(payload: &str) -> Result<UsageEvent> {
    let parts: Vec<&str> = payload.split('\t').collect();
    if parts.len() != 3 {
        bail!("invalid usage payload")
    }
    Ok(UsageEvent {
        input_tokens: parts[0].parse()?,
        output_tokens: parts[1].parse()?,
        cache_input_tokens: parts[2].parse()?,
    })
}
